File: src/ipc/layout.ts
import { invoke } from '@tauri-apps/api/core';

/**
 * Раскладка окна — дерево областей (Р-207).
 *
 * Приходит из ядра и в ядре меняется: всякая команда ниже возвращает дерево
 * целиком, и фронтенд заменяет им свою копию. «Что изменилось» не передаётся
 * нарочно — расхождение между копией и оригиналом так не накапливается.
 */

export type Direction = 'row' | 'column';

export interface PaneNode {
  kind: 'pane';
  id: number;
  /** Порядок в списке — порядок вкладок на полосе. */
  tabs: number[];
  active: number | null;
}

export interface SplitNode {
  kind: 'split';
  id: number;
  direction: Direction;
  /** Доля первого потомка. */
  ratio: number;
  first: LayoutNode;
  second: LayoutNode;
}

export type LayoutNode = PaneNode | SplitNode;

export interface Layout {
  root: LayoutNode;
  activePane: number;
  nextId: number;
}

export const layoutState = (): Promise<Layout> => invoke('layout_state');

export const setActivePane = (pane: number): Promise<Layout> =>
  invoke('set_active_pane', { pane });

export const setActiveTab = (pane: number, id: number): Promise<Layout> =>
  invoke('set_active_tab', { pane, id });

export const reorderTab = (pane: number, id: number, to: number): Promise<Layout> =>
  invoke('reorder_tab', { pane, id, to });

/** Убрать вкладку из одной области; буфер остаётся жить в других (Р-209). */
export const removeTab = (pane: number, id: number): Promise<Layout> =>
  invoke('remove_tab', { pane, id });

/** Разделить область. С буфером новая область получает его зеркало. */
export const splitPane = (
  pane: number,
  direction: Direction,
  id: number | null,
): Promise<Layout> => invoke('split_pane', { pane, direction, id });

export const closePane = (pane: number): Promise<Layout> => invoke('close_pane', { pane });

export const moveTab = (
  id: number,
  from: number,
  to: number,
  at: number | null,
): Promise<Layout> => invoke('move_tab', { id, from, to, at });

export const setSplitRatio = (split: number, ratio: number): Promise<Layout> =>
  invoke('set_split_ratio', { split, ratio });

File: src-tauri/src/commands/layout.rs
//! Команды раскладки окна — области редактора (Р-207).
//!
//! Логики здесь нет: дерево живёт в `model/layout.rs`, а команды только
//! берут блокировку, зовут одну операцию и отдают раскладку целиком.
//! Целиком, а не «что изменилось»: фронтенд заменяет свою копию ответом,
//! и расхождение между ним и ядром не накапливается.

use crate::model::buffer::BufferId;
use crate::model::layout::{Direction, Layout, NodeId};
use crate::state::AppState;

fn snapshot(state: &tauri::State<'_, AppState>) -> Layout {
    state
        .layout
        .lock()
        .expect("раскладка повреждена")
        .clone()
}

/// Раскладка как она есть.
#[tauri::command]
pub fn layout_state(state: tauri::State<'_, AppState>) -> Layout {
    snapshot(&state)
}

/// Фокус ушёл в область (Р-210).
#[tauri::command]
pub fn set_active_pane(state: tauri::State<'_, AppState>, pane: NodeId) -> Layout {
    let mut layout = state.layout.lock().expect("раскладка повреждена");
    layout.set_active_pane(pane);
    layout.clone()
}

/// Вкладку выбрали щелчком: она активна в своей области, область — в окне.
#[tauri::command]
pub fn set_active_tab(state: tauri::State<'_, AppState>, pane: NodeId, id: BufferId) -> Layout {
    let mut layout = state.layout.lock().expect("раскладка повреждена");
    layout.set_active_tab(pane, id);
    layout.clone()
}

/// Вкладку переставили внутри области.
#[tauri::command]
pub fn reorder_tab(
    state: tauri::State<'_, AppState>,
    pane: NodeId,
    id: BufferId,
    to: usize,
) -> Layout {
    let mut layout = state.layout.lock().expect("раскладка повреждена");
    layout.reorder(pane, id, to);
    layout.clone()
}

/// Убрать вкладку из одной области, оставив буфер жить в других (Р-209).
/// Из последней области вкладку так не убирают — это уже `close_buffer`.
#[tauri::command]
pub fn remove_tab(state: tauri::State<'_, AppState>, pane: NodeId, id: BufferId) -> Layout {
    let mut layout = state.layout.lock().expect("раскладка повреждена");
    if layout.panes_with(id).len() > 1 {
        layout.remove(pane, id);
    }
    layout.clone()
}

/// Разделить область. С буфером — новая область получает его вкладку
/// (зеркало, Р-209), без — остаётся пустой до первого открытия.
#[tauri::command]
pub fn split_pane(
    state: tauri::State<'_, AppState>,
    pane: NodeId,
    direction: Direction,
    id: Option<BufferId>,
) -> Layout {
    let mut layout = state.layout.lock().expect("раскладка повреждена");
    layout.split(pane, direction, id);
    layout.clone()
}

/// Закрыть область. Буферы, оставшиеся без области, закрывает фронтенд —
/// он же перед этим спрашивает про несохранённое.
#[tauri::command]
pub fn close_pane(state: tauri::State<'_, AppState>, pane: NodeId) -> Layout {
    let mut layout = state.layout.lock().expect("раскладка повреждена");
    layout.close_pane(pane);
    layout.clone()
}

/// Перенести вкладку между областями; внутри одной — переставить.
#[tauri::command]
pub fn move_tab(
    state: tauri::State<'_, AppState>,
    id: BufferId,
    from: NodeId,
    to: NodeId,
    at: Option<usize>,
) -> Layout {
    let mut layout = state.layout.lock().expect("раскладка повреждена");
    layout.move_tab(id, from, to, at);
    layout.clone()
}

/// Границу между областями подвинули.
#[tauri::command]
pub fn set_split_ratio(state: tauri::State<'_, AppState>, split: NodeId, ratio: f64) -> Layout {
    let mut layout = state.layout.lock().expect("раскладка повреждена");
    layout.set_ratio(split, ratio);
    layout.clone()
}
